src/L1/rollup/L1MessageQueue.sol

```diff
@@ -367,9 +373,53 @@ contract L1MessageQueue is OwnableUpgradeable, IL1MessageQueue {
         emit DequeueTransaction(_startIndex, _count, _skippedBitmap);
     }
 
+    /// @inheritdoc IL1MessageQueue
+    /// @dev Caller should make sure `_startIndex < pendingQueueIndex` to reduce unnecessary contract call.
+    function resetPoppedCrossDomainMessage(uint256 _startIndex) external override onlyScrollChain {
+        uint256 cachedPendingQueueIndex = pendingQueueIndex;
+        if (_startIndex == cachedPendingQueueIndex) return;
+
+        require(_startIndex >= nextUnfinalizedQueueIndex, "reset finalized messages");
+        require(_startIndex < cachedPendingQueueIndex, "reset pending messages");
+
+        unchecked {
+            uint256 count = cachedPendingQueueIndex - _startIndex;
+            uint256 bucket = _startIndex >> 8;
+            uint256 offset = _startIndex & 0xff;
+            skippedMessageBitmap[bucket] &= (1 << offset) - 1;
+            uint256 numResetMessages = 256 - offset;
+            while (numResetMessages < count) {
+                bucket += 1;
+                uint256 bitmap = skippedMessageBitmap[bucket];
+                if (bitmap > 0) skippedMessageBitmap[bucket] = 0;
+                numResetMessages += 256;
+            }
+        }
+
+        pendingQueueIndex = _startIndex;
+        emit ResetDequeuedTransaction(_startIndex);
+    }
+
+    /// @inheritdoc IL1MessageQueue
+    function finalizePoppedCrossDomainMessage(uint256 _newFinalizedQueueIndexPlusOne)
+        external
+        override
+        onlyScrollChain
+    {
+        uint256 cachedFinalizedQueueIndexPlusOne = nextUnfinalizedQueueIndex;
+        if (_newFinalizedQueueIndexPlusOne == cachedFinalizedQueueIndexPlusOne) return;
+        require(_newFinalizedQueueIndexPlusOne > cachedFinalizedQueueIndexPlusOne, "finalized index too small");
+        require(_newFinalizedQueueIndexPlusOne <= pendingQueueIndex, "finalized index too large");
+
+        nextUnfinalizedQueueIndex = _newFinalizedQueueIndexPlusOne;
+        unchecked {
+            emit FinalizedDequeuedTransaction(_newFinalizedQueueIndexPlusOne - 1);
```

Comment:
Currently bridge-history-backend only monitors `DequeueTransaction` and marks the skipped l1 messages immediately.

For UX of refunding, bridge-history-backend should monitor this event and tell Frontend that it can enable refunds after this event is emitted, otherwise, the tx will be reverted. 

One feasible change (without changing the Frontend) IMO:
1. adding a "pre-skipped" status, update l1 messages by listening to DequeueTransaction.
2. update status as "skipped" after FinalizedDequeuedTransaction (l1 messages <= _newFinalizedQueueIndexPlusOne - 1).

Do you think this is a real concern, or there are some other simple walkarounds? @zzq0826 @zimpha

Reply:
The bridge-history-backend should also monitor `ResetDequeuedTransaction`, in case batch reverting.

The frontend can check the `queueIndex` with `nextUnfinalizedQueueIndex` to decide whether it can enable refund.

Reply:
got it. will also add tracking `ResetDequeuedTransaction`, I think the backend can add a new status called "pre-skipped", for the frontend:
1. if the status is "pre-skipped", which means the message is very likely to be skipped (unless we revert the batch), but the user cannot trigger a refund yet.
2. if the status is "skipped", which means the message is skipped and the user can trigger the refund.
In this way, frontend does not need extra codes to check if the refund is available, only leaving the notification consideration to @zzq0826.

Reply:
but yeah. the method using "pre-skipped" won't be forward-compatible. @zimpha's solution is more reasonable and forward-compatible.